crates/storage/provider/src/providers/snapshot/manager.rs

```diff
@@ -39,30 +67,132 @@ impl SnapshotProvider {
+        // If we have a path, then get the block range and transaction range from its name.
+        // Otherwise, check `self.available_snapshots`
+        let snapshot_ranges = match path {
+            Some(path) => SnapshotSegment::parse_filename(
+                path.file_name().ok_or_else(|| ProviderError::MissingSnapshot)?,
+            )
+            .and_then(|(parsed_segment, block_range, tx_range)| {
+                if parsed_segment == segment {
+                    return Some((block_range, tx_range));
+                }
+                None
+            }),
+            None => fn_ranges(),
+        };
+
+        // Return cached `LoadedJar` or insert it for the first time, and then, return it.
+        match snapshot_ranges {
+            Some((block_range, tx_range)) => {
+                self.get_or_create_jar_provider(segment, &block_range, &tx_range)
+            }
+            None => Err(ProviderError::MissingSnapshot.into()),
         }
+    }
 
-        if let Some(path) = &path {
-            self.map.insert(key, LoadedJar::new(NippyJar::load(path)?)?);
+    /// Given a segment, block range and transaction range it returns a cached
+    /// [`SnapshotJarProvider`]. TODO: we should check the size and pop N if there's too many.
+    fn get_or_create_jar_provider(
+        &self,
+        segment: SnapshotSegment,
+        block_range: &RangeInclusive<u64>,
+        tx_range: &RangeInclusive<u64>,
+    ) -> Result<SnapshotJarProvider<'_>, reth_interfaces::RethError> {
+        let key = (*block_range.end(), segment);
+        if let Some(jar) = self.map.get(&key) {
+            Ok(jar.into())
         } else {
-            path = Some(self.path.join(segment.filename(
-                &((snapshot * BLOCKS_PER_SNAPSHOT)..=((snapshot + 1) * BLOCKS_PER_SNAPSHOT - 1)),
-            )));
+            self.map.insert(
+                key,
+                LoadedJar::new(NippyJar::load(
+                    &self.path.join(segment.filename(block_range, tx_range)),
+                )?)?,
+            );
+            Ok(self.map.get(&key).expect("qed").into())
+        }
+    }
+
+    /// Gets a snapshot segment's block range and transaction range from the provider inner block
+    /// index.
+    fn get_segment_ranges_from_block(
+        &self,
+        segment: SnapshotSegment,
+        block: u64,
+    ) -> Option<(RangeInclusive<BlockNumber>, RangeInclusive<TxNumber>)> {
+        let snapshots = self.snapshots_block_index.read();
+        let segment_snapshots = snapshots.get(&segment)?;
+
+        // It's more probable that the request comes from a newer block height, so we iterate
+        // the snapshots in reverse.
+        let mut snapshots_rev_iter = segment_snapshots.iter().rev().peekable();
+
+        while let Some((block_end, tx_range)) = snapshots_rev_iter.next() {
+            let block_start =
+                snapshots_rev_iter.peek().map(|(block_end, _)| *block_end + 1).unwrap_or(0);
```

Comment:
can we add a comment here why it's correct to do `unwrap_or(0)` and always return a range on latest snapshot?